### PR TITLE
array-processing-tool

### PR DESCRIPTION
**Pull request by romos21**
Added array processing tools tasks

### CODE REVIEW

File: assets/js/arrayProcessingTool.js [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+  console.log(object.arr);
+  if(object.arr.includes(NaN)){
+      resultHTMLElement.style.backgroundColor='#f85454'
+      resultHTMLElement.textContent='some element(s) in array is not correct. Result could fail';
+  }
+};
+
+const object={
+    arr:[],
+    subSum: function (){
```
Comment (ghost) [inline assets/js/arrayProcessingTool.js]: The program returns an invalid answer on the next input data:
- -1 2 3 -9
- -1 2 3 -9 11
- -2 -1 1 2
- 100 -9 2 -3 5
- 2 -1

Comment (ghost) [inline assets/js/arrayProcessingTool.js]: The program uses only one solution (O(n^2))

File: assets/js/arrayProcessingTool.js [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+                    currentSum += this.arr[i + j];
+                }
+            }
+            if (currentSum > sum) {
+                sum = currentSum;
+            }
+            return sum;
+        }
+    },
+    search:function (){
```
Comment (ghost) [inline assets/js/arrayProcessingTool.js]: The program returns an invalid median for the next input data:

- 2 3 1

Please, look at the definition of the median ([link](https://ru.wikipedia.org/wiki/%D0%9C%D0%B5%D0%B4%D0%B8%D0%B0%D0%BD%D0%B0_(%D1%81%D1%82%D0%B0%D1%82%D0%B8%D1%81%D1%82%D0%B8%D0%BA%D0%B0)) )

File: assets/js/arrayProcessingTool.js [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        for (let i=1;i<arrLength;i++){
+            if (this.arr[i]>max){
+                max=this.arr[i];
+            }
+            if (this.arr[i]<min){
+                min=this.arr[i];
+            }
+        }
+        const middle=this.arr[parseInt(arrLength/2)];
+        return 'max: '+max+' mai: '+min+' middle:'+middle;
```
Comment (ghost) [inline assets/js/arrayProcessingTool.js]: `return 'max: '+max+' min: '+min+' median:'+middle;`

File: arrayProcessingTool.html [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+<head>
+    <meta charset="UTF-8">
+    <link rel="stylesheet" href="assets/styles/arrayProcessingTool.css">
+    <title>Title</title>
+</head>
+<body>
+<span>Enter array in '2 3 4...' format</span>
+<input id="array">
+<br/>
+<button id="search-button">search</button>
```
Comment (ghost) [inline arrayProcessingTool.html:12]: _UI improvement_ : add the protection from submitting buttons when the user enters incorrect input data.

File: assets/js/arrayProcessingTool.js [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+inputArrayHTMLElement.onchange=()=>{
+  object.arr=inputArrayHTMLElement.value.split(' ').map(el=>Number(el));
+  console.log(object.arr);
+  if(object.arr.includes(NaN)){
+      resultHTMLElement.style.backgroundColor='#f85454'
+      resultHTMLElement.textContent='some element(s) in array is not correct. Result could fail';
+  }
+};
+
+const object={
```
Comment (ghost) [inline assets/js/arrayProcessingTool.js]: Give an understandable name for this variable

**Review by ghost — COMMENTED** [review]
@romos21 consider fixing these issues

File: assets/js/arrayProcessingTool.js (hunk 5 above)
Comment (romos21) [reply assets/js/arrayProcessingTool.js]: fixed

File: arrayProcessingTool.html (hunk 4 above)
Comment (romos21) [reply arrayProcessingTool.html:12]: fixed

File: assets/js/arrayProcessingTool.js (hunk 3 above)
Comment (romos21) [reply assets/js/arrayProcessingTool.js]: fixed

File: assets/js/arrayProcessingTool.js (hunk 2 above)
Comment (romos21) [reply assets/js/arrayProcessingTool.js]: fixed

File: assets/js/arrayProcessingTool.js (hunk 1 above)
Comment (romos21) [reply assets/js/arrayProcessingTool.js]: fixed

Comment (romos21) [reply assets/js/arrayProcessingTool.js]: fixed

File: assets/js/arrayProcessingTool.js [hunk 6]
```diff
@@ -1,94 +1,187 @@
...
+                if (j !== i) {
+                    if (arrTmp[j] < arrTmp[i]) {
+                        lessElementsLength++;
+                    } else {
+                        greatElementsLength++;
+                    }
+                }
+            }
+            if (lessElementsLength === greatElementsLength) {
+                mid = arrTmp[i];
```
Comment (ghost) [inline assets/js/arrayProcessingTool.js]: The program returns invalid median for the next input data:

- -1 2 3 -9 (median should be equal to 0.5)

- 11 9 3 5 5 (median should be equal to 5)

- 1 3 5 7 (median should be equal to 4)

File: assets/js/arrayProcessingTool.js [hunk 7]
```diff
@@ -1,94 +1,187 @@
...
+    },
+    negativeNumbersCheck: function (){
+        const arrTmpM = this.arr.filter(el => el < 0);
+        if (arrTmpM.length === this.arr.length) {
+            let result = this.search().split('').filter(el => Number(el));
+            return -result[0];
+        }
+        return null;
+    },
+    subSumOn1: function () {
```
Comment (ghost) [inline assets/js/arrayProcessingTool.js:24]: The program shows the invalid result with the next input data:

- -1 -2 -3 (should be 0)

- -1 2 3 -9 11 (should be 11)

File: assets/js/arrayProcessingTool.js [hunk 8]
```diff
@@ -1,94 +1,187 @@
...
+            if (currSum + arrTmp[i] >= 0) {
+                currSum += arrTmp[i];
+            }
+            if (currSum > maxSum) {
+                maxSum = currSum;
+            }
+        }
+        return maxSum;
+    },
+    subSumOn2: function () {
```
Comment (ghost) [inline assets/js/arrayProcessingTool.js:41]: The program shows the invalid result with the next input data:

- -1 -2 -3 (should be 0)

File: assets/js/arrayProcessingTool.js [hunk 9]
```diff
@@ -1,94 +1,187 @@
...
-            return {};
+            return 'array is empty';
         }
-        if (arrLength===1){
-            return {max:this.arr[0],min:this.arr[0],middle:this.arr[0]};
+        let max = arrTmp[0], min = arrTmp[0];
+        let canFindMiddle = arrLength % 2;
+        let mid = canFindMiddle ? arrTmp[0] : '';
+        if (arrLength === 1) {
+            return 'max: ' + arrTmp[0] + ' min: ' + arrTmp[0] + ' middle:' + arrTmp[0];
```
Comment (ghost) [inline assets/js/arrayProcessingTool.js]: Please, replace the word "middle" with "median".

File: assets/js/arrayProcessingTool.js [hunk 10]
```diff
@@ -1,94 +1,187 @@
...
         }
-        let max=this.arr[0],min=this.arr[0];
-        for (let i=1;i<arrLength;i++){
-            if (this.arr[i]>max){
-                max=this.arr[i];
+        if (arrLength === 2) {
+            if (arrTmp[1] > arrTmp[0]) {
+                return 'max: ' + arrTmp[1] + ' min: ' + arrTmp[0] + ' middle:' + mid;
```
Comment (ghost) [inline assets/js/arrayProcessingTool.js]: _Code style_: replace string formatting repetitive code. Define a function for it or make a single return in the method.

**Review by ghost — COMMENTED** [review]
@romos21 just a couple of fixes remain to add

File: assets/js/arrayProcessingTool.js (hunk 10 above)
Comment (romos21) [reply assets/js/arrayProcessingTool.js]: fixed

File: assets/js/arrayProcessingTool.js (hunk 9 above)
Comment (romos21) [reply assets/js/arrayProcessingTool.js]: fixed

File: assets/js/arrayProcessingTool.js (hunk 8 above)
Comment (romos21) [reply assets/js/arrayProcessingTool.js:41]: fixed

File: assets/js/arrayProcessingTool.js (hunk 7 above)
Comment (romos21) [reply assets/js/arrayProcessingTool.js:24]: fixed

File: assets/js/arrayProcessingTool.js (hunk 6 above)
Comment (romos21) [reply assets/js/arrayProcessingTool.js]: fixed

File: assets/js/arrayProcessingTool.js (hunk 5 above)
Comment (romos21) [reply assets/js/arrayProcessingTool.js]: fixed

**Comment by ghost** [comment]
@romos21 Everything is good

**Pull request merged**